lightning/src/onion_message/packet.rs

```diff
@@ -108,10 +109,8 @@ pub(super) enum Payload<T: CustomOnionMessageContents> {
 /// The contents of an onion message. In the context of offers, this would be the invoice, invoice
 /// request, or invoice error.
 pub enum OnionMessageContents<T: CustomOnionMessageContents> {
-	// Coming soon:
-	// Invoice,
-	// InvoiceRequest,
-	// InvoiceError,
+	/// A message related to BOLT 12 Offers.
+	Offers(OffersMessage),
```

Comment:
This was largely for code separation of each category with a corresponding handler. Offer messages in particular have the requirement of needing to keep around the `Vec<u8>` and dealing with `ParseError`s, so this behavior can be encapsulated in `OffersMessage`.

Reply:
> > This was largely for code separation of each category with a corresponding handler.
> 
> Should the same apply to P2P messages?
> 

I don't have a strong opinion, though dividing them by handler trait seems natural. Then again, having a separate method for handling each type in the trait may be preferable, which reduces the need for a wrapper enum.

> > Offer messages in particular have the requirement of needing to keep around the Vec and dealing with ParseErrors, so this behavior can be encapsulated in OffersMessage.
> 
> Isn't this handled at the message level? When we parse an offers message that message holds the Vec, there's no special handling at the messaging layer.

The offers messages don't implement `Readable`. They implement `TryFrom<Vec<u8>>` with `ParseError` (and thus `SemanticError`) as the error type. `OffersMessage` encapsulates this.

I can remove `OffersMessage` if you prefer. I guess it would remove the boilerplate from that file, leaving just the handler trait and `ChannelManager` implementation. Then the common parsing / error handling code can move to `packet.rs`.

Reply:
> Should this change now that they're being read from streams? Or should we have a totally different message reading trait for onion messages more generally?

I suppose we could implement `Readable` on them now that there's a new `DecodeError` variant to pass back the error in `String` form. But the `Vec<u8>` part is really only specific to offers messages. So I don't think there's a need to do anything different more generally for onion messages.

Might make sense to implement `Readable` on the applicable messages, though, that reads into a `Vec` first. I guess the argument against it is that we don't need to expose a `Readable` interface to the outside world, so should we not?

> I don't feel super strongly, no, I'm okay with a difference here, I just want to make sure we have a strong reason for it given its awkward we have two separate APIs that look different for similar concepts.

The less hierarchical version is a bit more compact, FWIW. So may go with it as it should also remove handler complexity.

```diff
diff --git a/lightning/src/onion_message/packet.rs b/lightning/src/onion_message/packet.rs
index 48a8fd393..7bdd7549e 100644
--- a/lightning/src/onion_message/packet.rs
+++ b/lightning/src/onion_message/packet.rs
@@ -15,12 +15,16 @@ use bitcoin::secp256k1::ecdh::SharedSecret;
 use crate::blinded_path::{BlindedPath, ForwardTlvs, ReceiveTlvs};
 use crate::ln::msgs::DecodeError;
 use crate::ln::onion_utils;
+use crate::offers::invoice::Invoice;
+use crate::offers::invoice_error::InvoiceError;
+use crate::offers::invoice_request::InvoiceRequest;
+use crate::offers::parse::ParseError;
 use super::messenger::CustomOnionMessageHandler;
-use super::offers::OffersMessage;
 use crate::util::chacha20poly1305rfc::{ChaChaPolyReadAdapter, ChaChaPolyWriteAdapter};
 use crate::util::ser::{BigSize, FixedLengthReader, LengthRead, LengthReadable, LengthReadableArgs, Readable, ReadableArgs, Writeable, Writer};
 
 use core::cmp;
+use core::convert::TryFrom;
 use crate::io::{self, Read};
 use crate::prelude::*;
 
@@ -105,12 +109,21 @@ pub(super) enum Payload<T: CustomOnionMessageContents> {
        }
 }
 
+// TLV record types for the `onionmsg_tlv` TLV stream as defined in BOLT 4.
+const INVOICE_REQUEST_TLV_TYPE: u64 = 64;
+const INVOICE_TLV_TYPE: u64 = 66;
+const INVOICE_ERROR_TLV_TYPE: u64 = 68;
+
 #[derive(Debug)]
 /// The contents of an onion message. In the context of offers, this would be the invoice, invoice
 /// request, or invoice error.
 pub enum OnionMessageContents<T: CustomOnionMessageContents> {
-       /// A message related to BOLT 12 Offers.
-       Offers(OffersMessage),
+       /// A BOLT 12 `invoice_request` message.
+       InvoiceRequest(InvoiceRequest),
+       /// A BOLT 12 `invoice` message.
+       Invoice(Invoice),
+       /// A BOLT 12 `invoice_error` message.
+       InvoiceError(InvoiceError),
        /// A custom onion message specified by the user.
        Custom(T),
 }
@@ -121,7 +134,9 @@ impl<T: CustomOnionMessageContents> OnionMessageContents<T> {
        /// This is not exported to bindings users as methods on non-cloneable enums are not currently exportable
        pub fn tlv_type(&self) -> u64 {
                match self {
-                       &OnionMessageContents::Offers(ref msg) => msg.tlv_type(),
+                       &OnionMessageContents::InvoiceRequest(_) => INVOICE_REQUEST_TLV_TYPE,
+                       &OnionMessageContents::Invoice(_) => INVOICE_TLV_TYPE,
+                       &OnionMessageContents::InvoiceError(_) => INVOICE_ERROR_TLV_TYPE,
                        &OnionMessageContents::Custom(ref msg) => msg.tlv_type(),
                }
        }
@@ -131,7 +146,9 @@ impl<T: CustomOnionMessageContents> OnionMessageContents<T> {
 impl<T: CustomOnionMessageContents> Writeable for OnionMessageContents<T> {
        fn write<W: Writer>(&self, w: &mut W) -> Result<(), io::Error> {
                match self {
-                       OnionMessageContents::Offers(msg) => Ok(msg.write(w)?),
+                       OnionMessageContents::InvoiceRequest(msg) => Ok(msg.write(w)?),
+                       OnionMessageContents::Invoice(msg) => Ok(msg.write(w)?),
+                       OnionMessageContents::InvoiceError(msg) => Ok(msg.write(w)?),
                        OnionMessageContents::Custom(msg) => Ok(msg.write(w)?),
                }
        }
@@ -223,14 +240,20 @@ impl<H: CustomOnionMessageHandler> ReadableArgs<(SharedSecret, &H)> for Payload<
 
                        message_type = Some(msg_type);
                        match msg_type {
-                               tlv_type if OffersMessage::is_known_type(tlv_type) => {
-                                       match OffersMessage::read(msg_reader, tlv_type) {
-                                               Ok(msg) => {
-                                                       message = Some(OnionMessageContents::Offers(msg));
-                                                       Ok(true)
-                                               },
-                                               Err(e) => Err(e),
-                                       }
+                               INVOICE_REQUEST_TLV_TYPE => {
+                                       let msg = parse_offers_message(msg_reader)?;
+                                       message = Some(OnionMessageContents::InvoiceRequest(msg));
+                                       Ok(true)
+                               },
+                               INVOICE_TLV_TYPE => {
+                                       let msg = parse_offers_message(msg_reader)?;
+                                       message = Some(OnionMessageContents::Invoice(msg));
+                                       Ok(true)
+                               },
+                               INVOICE_ERROR_TLV_TYPE => {
+                                       let msg = parse_offers_message(msg_reader)?;
+                                       message = Some(OnionMessageContents::InvoiceError(msg));
+                                       Ok(true)
                                },
                                _ => match handler.read_custom_message(msg_type, msg_reader) {
                                        Ok(Some(msg)) => {
@@ -264,6 +287,25 @@ impl<H: CustomOnionMessageHandler> ReadableArgs<(SharedSecret, &H)> for Payload<
        }
 }
 
+fn parse_offers_message<R: Read, T: TryFrom<Vec<u8>, Error = ParseError>>(
+       r: &mut R
+) -> Result<T, DecodeError> {
+       let mut bytes = Vec::new();
+       r.read_to_end(&mut bytes).unwrap();
+
+       match T::try_from(bytes) {
+               Ok(message) => Ok(message),
+               Err(ParseError::Decode(e)) => Err(e),
+               Err(ParseError::InvalidSemantics(e)) => {
+                       Err(DecodeError::InvalidSemantics(format!("{:?}", e)))
+               },
+               Err(ParseError::InvalidSignature(e)) => {
+                       Err(DecodeError::InvalidSemantics(format!("{:?}", e)))
+               },
+               Err(_) => Err(DecodeError::InvalidValue),
+       }
+}
+
 /// When reading a packet off the wire, we don't know a priori whether the packet is to be forwarded
 /// or received. Thus we read a ControlTlvs rather than reading a ForwardControlTlvs or
 /// ReceiveControlTlvs directly.
```

Reply:
Ended up sticking with `OffersMessage` after all. I went down the path of making individual handler methods with specific `Result` types (e.g., `Result<Invoice, InvoiceError>`), which seemed nice... until I needed to implement the trait for `IgnoringMessageHandler`. Would need to introduce an `Option` into the return types, which would be kinda ugly. At least handlers will be restricted to a specific category of messages this way.